tagstudio/src/qt/widgets/preview/field_containers.py

```diff
@@ -0,0 +1,520 @@
+            "}"
+        )
+        self.scroll_area.setWidget(scroll_container)
+
+        root_layout = QHBoxLayout(self)
+        root_layout.setContentsMargins(0, 0, 0, 0)
+        root_layout.addWidget(self.scroll_area)
+
+    def update_from_entry(self, entry_id: int, update_badges: bool = True):
+        """Update tags and fields from a single Entry source."""
+        logger.warning("[FieldContainers] Updating Selection", entry_id=entry_id)
+
+        self.cached_entries = [self.lib.get_entry_full(entry_id)]
+        entry_ = self.cached_entries[0]
+        container_len: int = len(entry_.fields)
+        container_index = 0
+
+        # Write tag container(s)
+        if entry_.tags:
+            categories = self.get_tag_categories(entry_.tags)
+            for cat, tags in sorted(categories.items(), key=lambda kv: (kv[0] is None, kv)):
+                self.write_tag_container(
+                    container_index, tags=tags, category_tag=cat, is_mixed=False
+                )
+                container_index += 1
+                container_len += 1
+        if update_badges:
+            self.emit_badge_signals({t.id for t in entry_.tags})
+
+        # Write field container(s)
+        for index, field in enumerate(entry_.fields, start=container_index):
+            self.write_container(index, field, is_mixed=False)
+
+        # Hide leftover container(s)
+        if len(self.containers) > container_len:
+            for i, c in enumerate(self.containers):
+                if i > (container_len - 1):
+                    c.setHidden(True)
+
+    def hide_containers(self):
+        """Hide all field and tag containers."""
+        for c in self.containers:
+            c.setHidden(True)
+
+    def get_tag_categories(self, tags: set[Tag]) -> dict[Tag | None, set[Tag]]:
+        """Get a dictionary of category tags mapped to their respective tags."""
+        cats: dict[Tag | None, set[Tag]] = {}
+        cats[None] = set()
+
+        base_tag_ids: set[int] = {x.id for x in tags}
+        exhausted: set[int] = set()
+        cluster_map: dict[int, set[int]] = {}
+
+        def add_to_cluster(tag_id: int, p_ids: list[int] | None = None):
+            """Maps a Tag's subtag's ID's back to it's parent Tag's ID.
+
+            Example:
+            Tag: ["Johnny Bravo", Parent Tags: "Cartoon Network (TV)", "Character"] maps to:
+            "Cartoon Network" -> Johnny Bravo,
+            "Character" -> "Johnny Bravo",
+            "TV" -> Johnny Bravo"
+            """
+            tag_obj = self.lib.get_tag(tag_id)  # Get full object
+            if p_ids is None:
+                p_ids = tag_obj.parent_ids
+
+            for p_id in p_ids:
+                if cluster_map.get(p_id) is None:
+                    cluster_map[p_id] = set()
+                # If the p_tag has p_tags of its own, recursively link those to the original Tag.
+                if tag.id not in cluster_map[p_id]:
+                    cluster_map[p_id].add(tag.id)
```

Comment:
there is a typo here: `tag.id` should be `tag_id` (or at least I can't find where `tag` was defined in this method)
```suggestion
                if tag_id not in cluster_map[p_id]:
                    cluster_map[p_id].add(tag_id)
```